app/pipes.go

```diff
@@ -0,0 +1,194 @@
+package app
+
+import (
+	"io"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/weaveworks/scope/common/mtime"
+	"github.com/weaveworks/scope/xfer"
+)
+
+const (
+	gcInterval  = 30 * time.Second // we check all the pipes every 30s
+	pipeTimeout = 1 * time.Minute  // pipes are closed when a client hasn't been connected for 1 minute
+	gcTimeout   = 1 * time.Minute  // after another 1 minute, tombstoned pipes are forgotten
+)
+
+// PipeRouter connects incoming and outgoing pipes.
+type PipeRouter struct {
+	sync.Mutex
+	wait  sync.WaitGroup
+	quit  chan struct{}
+	pipes map[string]*pipe
+}
+
+// for each end of the pipe, we keep a reference count & lastUsedTIme,
+// such that we can timeout pipes when either end is inactive.
+type end struct {
+	refCount     int
+	lastUsedTime time.Time
+}
+
+type pipe struct {
+	ui, probe     end
+	tombstoneTime time.Time
+
+	xfer.Pipe
+}
+
+// RegisterPipeRoutes registers the pipe routes
+func RegisterPipeRoutes(router *mux.Router) *PipeRouter {
```

Comment:
A little, yeah.  The returned pipe router is only used in tests TBH, to ensure the background goroutine shut sdown.  I don't think we need to worry about that in production though.